File: backend/app/schemas/draft_schedule.py
from datetime import date
from enum import StrEnum

from pydantic import BaseModel, ConfigDict, Field


class FailureCode(StrEnum):
    INSUFFICIENT_ROOM_CAPACITY = "INSUFFICIENT_ROOM_CAPACITY"
    INVALID_SESSION_PREFERENCE = "INVALID_SESSION_PREFERENCE"
    NO_FITTING_TIME_WINDOW = "NO_FITTING_TIME_WINDOW"
    INSUFFICIENT_SEMESTER_CAPACITY = "INSUFFICIENT_SEMESTER_CAPACITY"
    INVALID_PLANNING_PERIOD = "INVALID_PLANNING_PERIOD"
    INVALID_TEACHING_WINDOW = "INVALID_TEACHING_WINDOW"
    MISSING_TEACHING_WINDOW = "MISSING_TEACHING_WINDOW"
    NO_FEASIBLE_RESOURCE = "NO_FEASIBLE_RESOURCE"


class SessionEditFailureCode(StrEnum):
    INVALID_SESSION_DATE = "INVALID_SESSION_DATE"
    INVALID_SESSION_TIME_RANGE = "INVALID_SESSION_TIME_RANGE"
    INSUFFICIENT_ROOM_CAPACITY = "INSUFFICIENT_ROOM_CAPACITY"
    DUPLICATE_SESSION_DATE = "DUPLICATE_SESSION_DATE"
    LECTURER_INELIGIBLE = "LECTURER_INELIGIBLE"
    ROOM_INELIGIBLE = "ROOM_INELIGIBLE"
    LECTURER_UNAVAILABLE = "LECTURER_UNAVAILABLE"
    ROOM_UNAVAILABLE = "ROOM_UNAVAILABLE"


class ValidationAlertCode(StrEnum):
    LECTURER_OVERLAP = "LECTURER_OVERLAP"
    ROOM_OVERLAP = "ROOM_OVERLAP"
    COHORT_OVERLAP = "COHORT_OVERLAP"
    ROOM_CAPACITY = "ROOM_CAPACITY"
    GENERATION_CONSTRAINT_VIOLATION = "GENERATION_CONSTRAINT_VIOLATION"
    STUDY_TYPE_WINDOW_VIOLATION = "STUDY_TYPE_WINDOW_VIOLATION"
    VALIDATION_DATA_MISSING = "VALIDATION_DATA_MISSING"
    LECTURER_UNAVAILABLE = "LECTURER_UNAVAILABLE"
    ROOM_UNAVAILABLE = "ROOM_UNAVAILABLE"
    LECTURER_INELIGIBLE = "LECTURER_INELIGIBLE"
    ROOM_INELIGIBLE = "ROOM_INELIGIBLE"


class UpdateDraftSessionRequest(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    date: str
    start_time: str = Field(alias="startTime")
    end_time: str = Field(alias="endTime")
    lecturer_id: int | None = Field(default=None, alias="lecturerId")
    room_id: int = Field(alias="roomId")


class PlanningPeriodInput(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    start_date: date = Field(alias="startDate")
    end_date: date = Field(alias="endDate")


class AllowedTeachingWindowInput(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    weekday: int
    start_time: str = Field(alias="startTime")
    end_time: str = Field(alias="endTime")
    source_time_window_id: int | None = Field(default=None, alias="sourceTimeWindowId")


class GenerateDraftScheduleRequest(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    semester_id: int = Field(alias="semesterId")
    planning_period: PlanningPeriodInput = Field(alias="planningPeriod")
    allowed_teaching_windows: list[AllowedTeachingWindowInput] = Field(alias="allowedTeachingWindows")


class PlanningPeriodResponse(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    start_date: date = Field(alias="startDate")
    end_date: date = Field(alias="endDate")


class AllowedTeachingWindowResponse(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    weekday: int
    start_time: str = Field(alias="startTime")
    end_time: str = Field(alias="endTime")
    source_time_window_id: int | None = Field(default=None, alias="sourceTimeWindowId")


class GenerationConstraintsResponse(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    course_id: int = Field(alias="courseId")
    semester_id: int = Field(alias="semesterId")
    is_custom: bool = Field(alias="isCustom")
    revision: int | None = None
    planning_period: PlanningPeriodResponse = Field(alias="planningPeriod")
    allowed_teaching_windows: list[AllowedTeachingWindowResponse] = Field(alias="allowedTeachingWindows")


class RelatedSessionResponse(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    session_id: int = Field(alias="sessionId")
    draft_schedule_id: int = Field(alias="draftScheduleId")
    course_id: int = Field(alias="courseId")
    course_name: str = Field(alias="courseName")
    date: date
    start_time: str = Field(alias="startTime")
    end_time: str = Field(alias="endTime")
    cohort_name: str = Field(alias="cohortName")
    lecturer_name: str = Field(alias="lecturerName")
    room_name: str = Field(alias="roomName")


class ValidationAlertResponse(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    code: ValidationAlertCode
    message: str
    related_sessions: list[RelatedSessionResponse] = Field(alias="relatedSessions")


class DraftSessionResponse(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    id: int
    date: date
    start_time: str = Field(alias="startTime")
    end_time: str = Field(alias="endTime")
    units: int
    course_id: int = Field(alias="courseId")
    lecturer_id: int = Field(alias="lecturerId")
    lecturer_name: str = Field(alias="lecturerName")
    lecturer_reference_code: str = Field(alias="lecturerReferenceCode")
    cohort_id: int = Field(alias="cohortId")
    room_id: int = Field(alias="roomId")
    room_name: str = Field(alias="roomName")
    room_reference_code: str = Field(alias="roomReferenceCode")
    study_type_id: int = Field(alias="studyTypeId")
    time_window_id: int | None = Field(alias="timeWindowId")
    constraint_window_index: int = Field(alias="constraintWindowIndex")
    validation_alerts: list[ValidationAlertResponse] = Field(
        default_factory=list,
        alias="validationAlerts",
    )
    lecturer: "PlanningResourceResponse"
    room: "PlanningResourceResponse"


class PlanningEntityResponse(BaseModel):
    id: int
    name: str


class PlanningResourceResponse(PlanningEntityResponse):
    reference_code: str = Field(alias="referenceCode")


class DraftScheduleContextResponse(BaseModel):
    course: PlanningEntityResponse
    cohort: PlanningEntityResponse
    cohort_size: int = Field(alias="cohortSize")
    lecturer: PlanningEntityResponse
    room: PlanningEntityResponse
    study_type: PlanningEntityResponse = Field(alias="studyType")


class DraftScheduleResponse(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    draft_schedule_id: int = Field(alias="draftScheduleId")
    revision: int
    course_id: int = Field(alias="courseId")
    semester_id: int = Field(alias="semesterId")
    context: DraftScheduleContextResponse
    sessions: list[DraftSessionResponse]


class GenerationFailure(BaseModel):
    code: FailureCode
    message: str


class GenerationFailureResponse(BaseModel):
    errors: list[GenerationFailure]


class SessionEditFailure(BaseModel):
    code: SessionEditFailureCode
    message: str


class SessionEditFailureResponse(BaseModel):
    errors: list[SessionEditFailure]


class ErrorResponse(BaseModel):
    message: str

File: client/src/api/draftSchedule.ts
export type DraftSession = {
  id: number
  date: string
  startTime: string
  endTime: string
  units: number
  courseId: number
  lecturerId: number
  lecturerName: string
  lecturerReferenceCode: string
  cohortId: number
  roomId: number
  roomName: string
  roomReferenceCode: string
  studyTypeId: number
  timeWindowId: number | null
  constraintWindowIndex: number
  validationAlerts: ValidationAlert[]
  lecturer: PlanningResource
  room: PlanningResource
}

export type RelatedSession = {
  sessionId: number
  draftScheduleId: number
  courseId: number
  courseName: string
  date: string
  startTime: string
  endTime: string
  cohortName: string
  lecturerName: string
  roomName: string
}

export type ValidationAlert = {
  code: ValidationAlertCode
  message: string
  relatedSessions: RelatedSession[]
  holidayDate?: string | null
  holidayName?: string | null
}

export type ValidationAlertCode =
  | 'LECTURER_OVERLAP' | 'ROOM_OVERLAP' | 'COHORT_OVERLAP' | 'ROOM_CAPACITY'
  | 'GENERATION_CONSTRAINT_VIOLATION' | 'STUDY_TYPE_WINDOW_VIOLATION' | 'VALIDATION_DATA_MISSING'
  | 'LECTURER_UNAVAILABLE' | 'ROOM_UNAVAILABLE' | 'LECTURER_INELIGIBLE' | 'ROOM_INELIGIBLE'
  | 'INSTITUTION_HOLIDAY'

export type PlanningPeriod = {
  startDate: string
  endDate: string
}

export type AllowedTeachingWindow = {
  weekday: number
  startTime: string
  endTime: string
  sourceTimeWindowId?: number | null
}

export type GenerationConstraints = {
  courseId: number
  semesterId: number
  isCustom: boolean
  revision?: number | null
  planningPeriod: PlanningPeriod
  allowedTeachingWindows: AllowedTeachingWindow[]
}

export type PlanningEntity = {
  id: number
  name: string
}

export type PlanningResource = PlanningEntity & { referenceCode: string }

export type DraftScheduleContext = {
  course: PlanningEntity
  cohort: PlanningEntity
  cohortSize: number
  lecturer: PlanningEntity
  room: PlanningEntity
  studyType: PlanningEntity
}

export type DraftSchedule = {
  draftScheduleId: number
  revision: number
  courseId: number
  semesterId: number
  context: DraftScheduleContext
  sessions: DraftSession[]
}

export type ReviewFilters = {
  courseId?: number
  cohortId?: number
  lecturerId?: number
  roomId?: number
  studyTypeId?: number
}

export type ViewMode = 'list' | 'weekly'

export type GenerationFailure = {
  code: string
  message: string
  holidayDate?: string
  holidayName?: string
}

export type UpdateDraftSessionRequest = {
  date: string
  startTime: string
  endTime: string
  lecturerId: number
  roomId: number
}

export type SessionEditFailure = {
  code: string
  message: string
}

export type CourseSemesterProgress = {
  totalUnits: number
  scheduledUnits: number
  remainingUnits: number
}

export type DraftScheduleMutationResult = {
  courseId: number
  semesterId: number
  scheduledUnits: number
  remainingUnits: number
  draftSchedule: DraftSchedule | null
}

export type MutationFailure = {
  code: string
  message: string
  currentRevision?: number | null
}

export type CreateManualDraftSessionRequest = {
  semesterId: number
  date: string
  startTime: string
  endTime: string
  units: number
  roomId: number
}

const API_BASE = import.meta.env.VITE_API_BASE_URL ?? ''

export async function generateDraftSchedule(
  courseId: number,
  semesterId: number,
  planningPeriod: PlanningPeriod,
  allowedTeachingWindows: AllowedTeachingWindow[],
): Promise<DraftSchedule> {
  const response = await request(
    `${API_BASE}/api/courses/${courseId}/draft-schedule/generate`,
    {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ semesterId, planningPeriod, allowedTeachingWindows }),
    },
  )
  if (response.status === 422 || response.status === 409) {
    const payload = await response.json()
    throw payload.errors as GenerationFailure[]
  }
  if (!response.ok) {
    throw [{ code: 'REQUEST_FAILED', message: await response.text() }]
  }
  return response.json()
}

export async function getGenerationConstraints(
  courseId: number,
  semesterId: number,
): Promise<GenerationConstraints> {
  const response = await request(
    `${API_BASE}/api/courses/${courseId}/generation-constraints?semesterId=${semesterId}`,
  )
  if (!response.ok) {
    throw [{ code: 'REQUEST_FAILED', message: 'Could not load generation constraints.' }]
  }
  return response.json()
}

export async function clearGenerationConstraints(courseId: number, semesterId: number): Promise<void> {
  const response = await request(
    `${API_BASE}/api/courses/${courseId}/generation-constraints?semesterId=${semesterId}`,
    { method: 'DELETE' },
  )
  if (!response.ok) {
    throw [{ code: 'REQUEST_FAILED', message: 'Could not clear generation constraints.' }]
  }
}

export async function getDraftSchedule(courseId: number, semesterId: number): Promise<DraftSchedule> {
  const response = await request(
    `${API_BASE}/api/courses/${courseId}/draft-schedule?semesterId=${semesterId}`,
  )
  if (!response.ok) {
    throw [{ code: 'NOT_FOUND', message: 'No generated draft schedule exists.' }]
  }
  return response.json()
}

export async function getDraftSchedules(semesterId: number): Promise<DraftSchedule[]> {
  const response = await request(`${API_BASE}/api/draft-schedules?semesterId=${semesterId}`)
  if (!response.ok) {
    throw [{ code: 'REQUEST_FAILED', message: 'Could not load generated draft schedules.' }]
  }
  return response.json()
}

export async function updateDraftSession(
  sessionId: number,
  payload: UpdateDraftSessionRequest,
): Promise<DraftSchedule> {
  const response = await request(`${API_BASE}/api/draft-sessions/${sessionId}`, {
    method: 'PATCH',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(payload),
  })
  if (response.status === 422) {
    const body = await response.json()
    throw body.errors as SessionEditFailure[]
  }
  if (!response.ok) {
    throw [{ code: 'REQUEST_FAILED', message: await response.text() }]
  }
  return response.json()
}

export async function createManualDraftSession(
  courseId: number,
  payload: CreateManualDraftSessionRequest,
): Promise<DraftScheduleMutationResult> {
  const response = await request(`${API_BASE}/api/courses/${courseId}/draft-schedule/sessions`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(payload),
  })
  return parseMutationResponse(response)
}

export async function deleteDraftSession(
  sessionId: number,
  expectedDraftScheduleId: number,
  expectedDraftRevision: number,
): Promise<DraftScheduleMutationResult> {
  const query = new URLSearchParams({
    expectedDraftScheduleId: String(expectedDraftScheduleId),
    expectedDraftRevision: String(expectedDraftRevision),
  })
  const response = await request(`${API_BASE}/api/draft-sessions/${sessionId}?${query}`, { method: 'DELETE' })
  return parseMutationResponse(response)
}

export async function clearCourseDraft(
  courseId: number,
  semesterId: number,
  expectedDraftScheduleId: number,
  expectedDraftRevision: number,
): Promise<DraftScheduleMutationResult> {
  const query = new URLSearchParams({
    semesterId: String(semesterId),
    expectedDraftScheduleId: String(expectedDraftScheduleId),
    expectedDraftRevision: String(expectedDraftRevision),
  })
  const response = await request(`${API_BASE}/api/courses/${courseId}/draft-schedule?${query}`, { method: 'DELETE' })
  return parseMutationResponse(response)
}

async function parseMutationResponse(response: Response): Promise<DraftScheduleMutationResult> {
  if (response.ok) return response.json()
  let payload: { errors?: MutationFailure[]; detail?: string } = {}
  try {
    payload = await response.json()
  } catch {
    // Use the stable fallback below when the backend did not return JSON.
  }
  if (payload.errors?.length) throw payload.errors
  throw [{
    code: response.status === 404 ? 'NOT_FOUND' : 'REQUEST_FAILED',
    message: payload.detail ?? 'The requested Draft Schedule change could not be completed.',
  }] satisfies MutationFailure[]
}

async function request(input: RequestInfo | URL, init?: RequestInit): Promise<Response> {
  try {
    return await fetch(input, init)
  } catch {
    throw [
      {
        code: 'NETWORK_ERROR',
        message: 'Could not reach the backend API. Check that FastAPI is running and CORS is enabled.',
      },
    ]
  }
}
